Support polygon-based subsetting in grdcut -F with GeoDataFrame

kshitijaucharmal opened the pull request:
**Description of proposed changes**

This PR enhances the pygmt.grdcut function to support polygon-based subsetting of grids using the -F option. Users can now provide polygons in three different formats:

- A geopandas.GeoDataFrame
- A shapely.geometry.Polygon
- A GMT ASCII polygon file

Additionally, two new boolean parameters, `crop` and `invert`, have been added to control how the polygon is applied:
- `crop`=True crops the output grid region to the bounding box of the polygon.
- `invert`=True inverts the selection, setting all nodes inside the polygon to NaN.

> I have added both of these as parameters to grdcut, but they only apply to polygon mode. Please let me know if this is undesirable, and how would I implement this otherwise

A gallery example has been added to demonstrate all three polygon input methods and the effects of the crop and invert options, showing original, cropped, inverted, and cropped+inverted grids. This improves PyGMT’s functionality for extracting regions of interest from raster grids and makes the workflow more Pythonic and flexible.

Fixes #1529 (implements rather)

<!-- If significant changes to the documentation are made, please insert the link to the documentation page after it has been built. -->
**Preview**: https://pygmt-dev--4135.org.readthedocs.build/en/4135/gallery/images/grdpolygon.html

Users can crop grids using polygons without manually creating ASCII files.
Example figure showing original grid and cropped and inverted grids

<img width="1871" height="405" alt="image" src="https://github.com/user-attachments/assets/7144fdd9-4e29-4587-a4a1-a5771ab1c579" />


**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [x] If adding new functionality, add an example to docstrings or tutorials.

kshitijaucharmal committed:
Add polygon-based subregion support to grdcut with gallery example

kshitijaucharmal committed:
Add crop and invert boolean support for polygon input

weiji14 reviewed (commented):
Thanks @kshitijaucharmal! Just a few comments for now. If you're able to add some unit tests to `test_grdcut.py`, that would be fantastic too!

kshitijaucharmal committed:
grdcut: PathLike for polygon input and improve example

- Checking for PathLike instead of str, bytes
- Inverted logic and comments to make it clearer
- Update gallery example:
  - removed "complex" from polygon desc.
  - improved docs
  - Switch CRS to OGC:CRS84

kshitijaucharmal committed:
Add polygon input support tests for grdcut

- Validated polygon input types, raising GMTValueError for unsupported types.
- Unit tests added for:
    - Shapely Polygon input
    - GeoDataFrame polygon input
    - GMT ASCII polygon file input
    - Crop (+c) and invert (+i) modifiers
    - Invalid polygon input type to raise error

kshitijaucharmal commented:
@weiji14 I have resolved all comments and added the unit tests as well, let me know if there are any more corrections! :)

weiji14 commented on pygmt/tests/test_grdcut.py:
```diff
@@ -2,13 +2,26 @@
 from pygmt.helpers.testing import load_static_earth_relief
+from shapely.geometry import Polygon
+
+# Reusable polygon geometry
+poly = Polygon(
+    [
+        (-52.5, -19.5),
+        (-51.5, -19.0),
+        (-50.5, -18.5),
+        (-51.5, -18.0),
+        (-52.5, -19.5),
+    ]
+)
```
This should be turned into a pytest fixture, see examples a few lines below.

weiji14 commented on pygmt/tests/test_grdcut.py:
```diff
@@ -78,3 +91,58 @@ def test_grdcut_invalid_kind(grid, region):
+    outgrid = grdcut(grid=grid, polygon=poly)
+    assert outgrid is not None
+    assert outgrid.rio.crs is None or str(outgrid.rio.crs) == ""  # CRS not enforced
+    assert outgrid.size > 0
```
Could you add a more precise assertion(s)? Ideally, you want to check the exact pixel count, and maybe also check that there are NaN values in the four corners or something.

kshitijaucharmal replied:
I understand, for the polygon each vertex will need to not be NaN

weiji14 commented on pygmt/src/grdcut.py:
```diff
@@ -86,6 +90,17 @@ def grdcut(
+    polygon : str, geopandas.GeoDataFrame, or shapely.geometry
+        Extract a subregion using a polygon. Can be either:
+        - A GMT ASCII polygon file (`.gmt`)
+        - A geopandas.GeoDataFrame (must have CRS EPSG:4326)
```
I'm not sure if that specific CRS is necessary, of if a CRS is needed actually. Other input types (ASCII file, shapely.geometry.Polygon) don't even have a CRS.

weiji14 commented on pygmt/src/grdcut.py:
```diff
@@ -86,6 +90,17 @@ def grdcut(
+        Extract a subregion using a polygon. Can be either:
+        - A GMT ASCII polygon file (`.gmt`)
+        - A geopandas.GeoDataFrame (must have CRS EPSG:4326)
+        - A shapely.geometry.Polygon or MultiPolygon
```
Could you add a unit test to confirm that MultiPolygon does work?

weiji14 reviewed (commented):
Thanks for adding the unit tests! Here are some more comments, and apologies for the failing CI tests, we have some issues with DVC running on forks that we'll probably try to sort out next week.

kshitijaucharmal committed:
Add polygon fixture and more precise grdcut unit tests

- Reusable polygon fixture for grdcut tests
- Improve assertions to check pixel counts and presence of NaNs outside polygons
- Add a test for MultiPolygon support

kshitijaucharmal commented:
I do not quite understand the precise assertions part, is checking for some pixels inside polygon to not be NaN enough? I hope the new cases are correct or at least in the right direction. Thanks !

kshitijaucharmal committed:
Merge branch 'main' into grdcut-polygon

weiji14 committed:
Merge branch 'main' into grdcut-polygon

kshitijaucharmal commented:
Hey @weiji14 are the CI/CD issues not fixed yet? Please let me know if I can help in any way, thanks !

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -130,12 +145,33 @@ def grdcut(
+                # assuming its geojson
+                elif hasattr(polygon_input, "__geo_interface__"):
+                    tmpfile = stack.enter_context(tempfile_from_geojson(polygon_input))
+                    aliasdict["F"] = tmpfile + modifiers
```
```suggestion
                # if file path provided or a geojson object
                if isinstance(polygon_input, PathLike) or hasattr(
                    polygon_input, "__geo_interface__"
                ):
                    aliasdict["F"] = str(vinpoly) + modifiers
```

seisman commented:
> Hey @weiji14 are the CI/CD issues not fixed yet? Please let me know if I can help in any way, thanks !

We still have issues with CI, but you can just go ahead to improve this PR. We will run the tests locally as a temporary workaround.

kshitijaucharmal committed:
Merge branch 'GenericMappingTools:main' into grdcut-polygon

kshitijaucharmal committed:
removed ExitStack

kshitijaucharmal commented:
Have made the changes requested, please check

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -14,14 +15,15 @@
 
 @fmt_docstring
-@use_alias(N="extend", S="circ_subregion", Z="z_subregion", f="coltypes")
+@use_alias(N="extend", S="circ_subregion", Z="z_subregion", f="coltypes", F="polygon")
```
We have implemented a new alias system and will get rid of the `use_alias` decorator in the future.
```suggestion
@use_alias(N="extend", S="circ_subregion", Z="z_subregion", f="coltypes")
```

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -133,9 +148,31 @@ def grdcut(
         with (
             lib.virtualfile_in(check_kind="raster", data=grid) as vingrd,
+            lib.virtualfile_in(
+                check_kind="vector", data=kwargs.get("F"), required=False
```
```suggestion
                check_kind="vector", data=polygon, required=False
```

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -133,9 +148,31 @@ def grdcut(
+                ):
+                    aliasdict["F"] = str(vinpoly) + modifiers
+                else:
+                    polygon_type = type(polygon_input).__name__
```
```suggestion
                    polygon_type = type(polygon).__name__
```

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -30,6 +32,8 @@ def grdcut(
     region: Sequence[float | str] | str | None = None,
     verbose: Literal["quiet", "error", "warning", "timing", "info", "compat", "debug"]
     | bool = False,
```
```suggestion
    | bool = False,
    polygon: PathLike | None = None,
```

seisman reviewed (commented):
I've made some suggestions, mainly to migrate the `polgyon` parameter to the new alias system

kshitijaucharmal committed:
migrate polygon param to new alias system

kshitijaucharmal commented:
Have made the changes, the new alias system is much better honestly :)

seisman commented on pygmt/src/grdcut.py:
```diff
@@ -133,9 +154,27 @@ def grdcut(
+                        err_msg,
+                        polygon_type,
+                        "Must be a PathLike, GeoDataFrame, or Shapely geometry.",
+                    )
```
This check can be done by the `data_kind` function, e.g.,
```
if polyon is not None and data_kind(polygon) not in ("file", "geojson"):
    raise GMTTypeError(...)
```
Ideally we should raise the error as early as possible, so please put it at line 140.

seisman replied:
Please see my comment above. The check can be done earlier.
```suggestion
            if polygon:
                aliasdict["F"] = str(vinpoly) + aliasdict.get("F", "")
```

seisman commented:
Kindly ping @kshitijaucharmal.

seisman commented:
Ping @kshitijaucharmal.

seisman committed:
Merge branch 'main' into grdcut-polygon

seisman committed:
Apply suggestions from code review

Co-authored-by: Dongdong Tian <seisman.info@gmail.com>

seisman committed:
Fix style checks

seisman committed:
Raise error as early as possible
